Sync data access content (#623)
aerogear/mobile-docs: modules/ROOT/pages/_partials/sync/server-data-access.adoc

= Data Access

> It is very likely a {sync-service} application will need to access data from a database for it to work. The GraphQL spec or Voyager libraries does not require an application to use a particular type of database, and the developer of the application can choose a database that is most suitable for the application. 

Reviewer: Is the point here that:

Typically when working with GraphQL, you use a database  to store data, for example, MongoDB or PostgreSQL.
To summarize the tasks associated with using a database from {sync-service}:

Author: kind of. My main point here is that the develop can choose whatever database that's suits their app.

Reviewer: does mentioning Voyager in this context introduce the possibility that it doesn't work with sync service, ie will users care that it works with Voyager?

Author: I don't think so. I think users do care, because some sync libraries work with only some type of database. For example, the old sync only works with mongodb. CouchDB only works with CouchDB etc. But with Voyager, you can use any.

From a very high level, here are the steps to access data from a database in a {sync-service} application:

. Initialize the database connection on app start
. Add the database connection to the `context` object
. Retrieve/persist data to the database in the application's resolvers

== Initialize Database Connection

Generally speaking, the application should set up the database connection on start, and fail if the connection can not be established. In order to connect to the database, usually the information like the hostname and port number of the database server, the credentials to authenticate etc will be required. It is recommended to use environment variables to set these details. 

> If the application is going to be deployed on OpenShift as described in the <<Running {sync-service} on OpenShift>> section, then the following environment variables will be setup for the application automatically:

Reviewer: so, these env vars aren't available when doing local dev?
Would it be typical that user *doesn't* hook up voyager server to db during local dev times?

Author: no, these env vars won't be available locally unless the developer set them explicitly. 

No, that's not typical. Usually devs do hook up to db during local dev.

Reviewer: so, i think the text is ambiguous here, what's the best way for me to propose edit, after merging this PR?

Author: create a new PR after this is merged?

[options="header"]
|====
|Field Name|Environment Variable Name
|Database Server Hostname|DB_HOSTNAME
|Database Server Port|DB_PORT
|Database Server Username|DB_USERNAME
|Database Server Password|DB_PASSWORD
|Database name|DB_NAME
|Use SSL|DB_SSL
|====

It is highly recommended that the application use these environment variables to setup the database connection.

> For example, an application needs to connect to a PostgreSQL database. The database, table and schema is already setup. On app start, the database connection can be established like this:

Reviewer 2: ```suggestion
For example, if an application needs to connect to a PostgresQL database (where all tables and schemas are already set up) then a database connection can be established on app start as follows:
```

[source,javascript]
----
//knex is a SQL query builder
const knex = require('knex')

const dbConfig = {
  database: process.env.DB_NAME || 'users',
  user: process.env.DB_USERNAME || 'postgresql',
  password: process.env.DB_PASSWORD || 'postgres',
  host: process.env.DB_HOSTNAME || '127.0.0.1',
  port: process.env.DB_PORT || '5432'
}

const db = knex({
  client: 'pg',
  connection: dbConfig
})
----

== Add Database Connection to `context` object

> Once the database connection is established, it needs to be accessible from a resolver so that it can be used to retrieve or persist data. The recommended way to achieve this is to add the database connection to the `context` object. 

Reviewer: kinda think this para is unnecessary

Author: why?

Reviewer: or maybe reword, something like:

Add the database connection to the `context` object so that a resolver can access the database.

?

Author: sure
